### DISCUSSION TITLE
Clipboard on Safari/iOS help request

### DISCUSSION THREAD

**Discussion by miketwenty1** [wasm-bindgen/wasm-bindgen#3781] [Q&A]
### Summary

Following up with a [discussion](https://github.com/rustwasm/wasm-bindgen/discussions/3749) here to get more visibility. 
May be related to: https://github.com/emilk/egui/issues/3480 ?

I'm trying to get clipboard working with Safari browsers / iOS. Any help would be greatly appreciated.

```rust
let _task = spawn_local(async move {
    let window = web_sys::window().expect("window");
    let nav = window.navigator().clipboard();
    match nav {
        Some(a) => {
            let p = a.write_text("this text copies to clipboard on android and desktop but not Safari/iOS");
            let result = wasm_bindgen_futures::JsFuture::from(p).await;
            match result {
                Ok(_) => info!("copy worked"),
                Err(e) => info!("copy fail {:?}", e),
            }
        }
        None => {
            warn!("failed to clippy the clipboard clip clip");
        }
    };
});
```

The error is `JsValue(NotAllowedError: The request is not allowed by the user agent or the platform in the current context, possibly because the user denied permission.`

The same error happens on both safari/macbook and safari/iOS.

**Comment by daxpedda** [comment]
Reading the documentation on [MDN](https://developer.mozilla.org/en-US/docs/Web/API/Clipboard/writeText) or [Caniuse](https://caniuse.com/mdn-api_clipboard_writetext), it seems that calling `Clipboard.writeText` requires some level of [transient activation](https://developer.mozilla.org/en-US/docs/Glossary/Transient_activation) (it specifically says `pointerdown` or `pointerup`, but I assume any transient activation should do?).

**Reply by miketwenty1** [reply]
Well, fwiw it works like a charm on literally everything I've tried except iOS (Because Safari) and Macbook Safari.
Anything you can suggest I try to help debug or progress/work around? The activation method I'm trying is interaction press with a ui button via bevy.

**Reply by daxpedda** [reply]
Reacting to a button press sounds about right, should be the `pointerdown` event. So unfortunately I have no clue what could be wrong.
Maybe Bevy is reacting to the button press in the next frame, so transient activation might have run out until then.

Just to make sure you could check if [`UserActivation.isActive`](https://developer.mozilla.org/en-US/docs/Web/API/UserActivation/isActive) is `true` right before you call `Clipboard.writeText`.

**Reply by miketwenty1** [reply]
seems `UserActivation` is a fairly new feature? I see it for 0.3.67 but not for 0.3.64. Bevy crates seem to be pegged to 0.3.64 and I'm really struggling to patch in 0.3.67. Any advice here for my cargo toml? Or is there a way to do this with 0.3.64? 

v0.3.64 seems to be the git tag of 0.2.87 which is a little bit confusing.
https://github.com/rustwasm/wasm-bindgen/tree/0.2.87/crates/web-sys.

Also, I think you're right about bevy maybe doing this in the next frame.. perhaps not an issue for android and chrome but an issue for safari.

I was trying to patch like this to no avail, I couldn't bring back UserActivation struct for some reason:
```toml
[patch.crates-io]
web-sys = { git = "https://github.com/rustwasm/wasm-bindgen.git", branch = "main", features = [
    "UserActivation",
] }
```

**Reply by daxpedda** [reply]
You can just add your own bindings locally:
```rust
#[wasm_bindgen]
extern "C" {
    type NavigatorExt;

    #[wasm_bindgen(method, getter, js_name = userActivation)]
    fn user_activation(this: &NavigatorExt) -> UserActivation;

    type UserActivation;

    #[wasm_bindgen(method, getter, js_name = isActive)]
    fn is_active(this: &UserActivation) -> bool;
}

let navigator: NavigatorExt = web_sys::window().unwrap().navigator().unchecked_into();
let is_active: bool = navigator.user_activation().is_active();
```

**Reply by miketwenty1** [reply]
You can see below: `is_active? true`

```
[Log] INFO src/.../../systems.rs:46 is_active? true (bevygame.js, line 495)

[Log] INFO src/.../../systems.rs:55 clipboard fail JsValue(NotAllowedError: The request is not allowed by the user agent or the platform in the current context, possibly because the user denied permission. (bevygame.js, line 495)
writeText@[native code]
@https://example-replace-text.com/client/out/bevygame.js:443:46
wasm-stub@[wasm code]
<?>.wasm-function[1284]@[wasm code]
<?>.wasm-function[5024]@[wasm code]
<?>.wasm-function[25422]@[wasm code]
wasm-stub@[wasm code]
__wbg_adapter_34@https://example-replace-text.com/client/out/bevygame.js:218:73
real@https://example-replace-text.com/client/out/bevygame.js:203:21)

```
<img width="505" alt="Screenshot 2024-01-14 at 4 46 02 PM" src="https://github.com/rustwasm/wasm-bindgen/assets/7761473/ac69f6a1-7dd6-44f2-b583-6272ed086607">

So I have a `info!()` showing the is_active. It's active for safari even though right after I'll get the permission denied.
If you have anything else for me to check, happy to gather more info.

**Reply by daxpedda** [reply]
By any chance are you maybe not in a [secure context](https://developer.mozilla.org/en-US/docs/Web/Security/Secure_Contexts)?

**Reply by miketwenty1** [reply]
logging out `window.is_secure_context();` as well now.
It's coming back as true. 

Anything left to try? lol

**Reply by daxpedda** [reply]
Yeah, maybe the requirement is really not a transient activation but you actually have to call it from the `pointerdown` or `pointerup` event callbacks (there will probably be others).

Could you try the following:
```js
window.addEventListener('pointerdown', async () => {
	console.info('`pointerdown` event')
	await navigator.clipboard.writeText('clipboard test')
})
```
Then just click in your window and see if you also get a `NotAllowedError`.

**Reply by miketwenty1** [reply]
This worked, the clipboard while on safari is now being populated with "clipboard test". If you have another instruction for me, I'll take it.

Edit: Or would it make sense to bypass bevy and somehow write the data out bevy to be copied?
Also note: For chrome it's still getting the data I need, so I assume it's copying to clipboard twice and overwriting "clipboard test" with the bevy data.

**Reply by daxpedda** [reply]
The problem is that Bevy reacts to button presses through events that are processed "later", which isn't inside the `pointerdown` callback anymore. So I don't think you can circumvent Bevy unless you implement your own `pointerdown` handler that also finds your button and everything connected to that.

There's nothing much you could do at this point except hoping for Safari to be more spec compliant in the future. I couldn't specifically find a bug report for this, but you could file a request to change `Clipboard.writeText()` to require a transient activation instead of these events. The only thing I could find is [this report](https://bugs.webkit.org/show_bug.cgi?id=222262), which might give Bevy a way to circumvent this mess after all if it ever gets resolved (though in an incredibly complicated way).

Alternatively you could still try [`Clipboard.write()`](https://developer.mozilla.org/en-US/docs/Web/API/Clipboard/write)? But keep in mind that it isn't supported by Firefox, so you might wanna do some funny browser detection magic around that.

**Reply by miketwenty1** [reply]
while trying a Clipboard.write()
I'm getting a different error:
`JsValue(TypeError: Value is not a sequence`

I'm doing:
```rust
let clipboard = window.navigator().clipboard().unwrap();
let jsval_data = JsValue::from_str(&string_data);
let p2 = clipboard.write(&jsval_data);
let result2 = wasm_bindgen_futures::JsFuture::from(p2).await;
```

Should I be doing something different?

**Reply by daxpedda** [reply]
It's a bit involved:
```rust
#[wasm_bindgen]
extern "C" {
    type ClipboardItem;

    #[wasm_bindgen(constructor)]
    fn new(data: &Object) -> ClipboardItem;
}

let clipboard = web_sys::window().unwrap().navigator().clipboard().unwrap();
let item_data = Object::new();
let item_value = Blob::new_with_blob_sequence_and_options(
    &Array::of1(&"clipboard test".into()),
    BlobPropertyBag::new().type_("text/plain"),
)
.unwrap();
Reflect::set(&item_data, &"text/plain".into(), &item_value).unwrap();
let item = ClipboardItem::new(&item_data);
let promise = clipboard.write(&Array::of1(&item));
```
That worked in Chrome.

The spec actually doesn't require a `Blob`, a `string` should have worked as well, but that's a limitation in Chrome apparently. If this doesn't work for you on Safari, it's probably best if you play around with it in JS first before trying to translate it into Rust `web-sys`.

FWIW Whis is the equivalent in JS:
```js
const item_value = new Blob(['clipboard test'], { type: 'text/plain' })
const item_data = { 'text/plain': item_value }
const item = new ClipboardItem(item_data)
const promise = navigator.clipboard.write([item])
```

**Reply by miketwenty1** [reply]
This now works on safari desktop! Essentially if the first method fails, I'll try a second clipboard method, so it doesn't mess up any prior functionality.

But this still doesn't work on iOS. I may have missed it before but I looked at the javascript console debugger for iOS.

The thing we setup before shows up every time I press the screen while the game is loaded. 
```js
window.addEventListener('pointerdown', async () => {
	console.info('`pointerdown` event')
	await navigator.clipboard.writeText('clipboard test')
})
```
<img width="1058" alt="Screenshot 2024-01-15 at 11 33 21 AM" src="https://github.com/rustwasm/wasm-bindgen/assets/7761473/48619ab6-6ded-4091-abd3-fc0cf98027a9">

`Unhandled Promise Rejection: NotAllowedError: The request is not allowed by the user agent or the platform in the current context, possibly because the user denied permission.`


Are there ideas of working around this last hurdle?

**Reply by daxpedda** [reply]
Could you try using the `Clipboard.write()` in `pointerdown` instead?
Are we also on a secure context on iOS as well?

**Reply by miketwenty1** [reply]
I've tried write() and it ends up with the original error: `The request is not allowed by the user agent or the platform in the current context, possibly because the user denied permission.`

I've given up on iOS with a traditional in bevy approach. 
I'm now attempting to just use wasm bindgen within bevy to manipulate the dom:

```html
<p id="textToCopy">default text</p>
<button id="copyButton" onclick="copyToClipboard()">Copy text</button>
```

I ideally want to trigger an event and set text of the `<p>` like a `document.getElementById("textToCopy").innerText = "Some String from bevy";`  to a given string and then do a `document.getElementById("copyButton").style.display = "block";` 
and then after another event do a `document.getElementById("copyButton").style.display = "hidden";`

Still trying to crunch through the wasm bindgen code for it. Open to receiving more `#[wasm_bindgen]` magic.

**Reply by daxpedda** [reply]
This should be pretty straight-forward with [`HtmlElement::style()`](https://docs.rs/web-sys/0.3.67/web_sys/struct.HtmlElement.html#method.style) and [`CssStyleDeclaration::set_property()`](https://docs.rs/web-sys/0.3.67/web_sys/struct.CssStyleDeclaration.html#method.set_property), no custom bindings necessary.

Let me know if you still have any issues.
Honestly would have been nice to figure out all the iOS stuff as well.

**Comment by eblack-leaf** [comment]
![image](https://github.com/rustwasm/wasm-bindgen/assets/113637960/e4d03ea0-862a-4388-bfb2-d6e0a22a2dca)

from [web.dev](https://web.dev/articles/async-clipboard)

I am stuck at iOS copy/paste but am not sure how to formulate the above info to a working example,
but it seems the user-activation could possibly be the reason for not having it read.

**Reply by daxpedda** [reply]
This WebKit bug is quite interesting, if this is correct it would explain why @miketwenty1 had such issues.
@miketwenty1 if you are still interested you should try if this solves your problem, e.g. transforming your last example to:
```js
const item_value = new Blob(['clipboard test'], { type: 'text/plain' })
const item_data = { 'text/plain': new Promise(async (resolve) => resolve(item_value)) }
const item = new ClipboardItem(item_data)
const promise = navigator.clipboard.write([item])
```

---

@eblack-leaf I'm happy to help.
You could start by simply trying if the following code works on iOS:
```
window.addEventListener('click', async () => {
    navigator.clipboard.write([
        new ClipboardItem({
            'text/plain': new Promise(async (resolve) => resolve(new Blob(['clipboard test'], { type: 'text/plain' }))),
        }),
    ])
})
```
To simplify this we should also try if this works without `async` entirely:
```
window.addEventListener('click', () => {
    navigator.clipboard.write([
        new ClipboardItem({
            'text/plain': new Promise((resolve) => resolve(new Blob(['clipboard test'], { type: 'text/plain' }))),
        }),
    ])
})
```
This should also work in every browser as far as I have tested.
Also try the `pointerdown` event or whatever event you want to actually use here.

Converting this to Rust should then be quite straightforward.

**Reply by eblack-leaf** [reply]
I am having trouble converting to `web_sys`. `ClipboardItem` does not have a constructor I can use in `web_sys` as far as I can understand. I can use a 'copy' event callback when I initialize the Clipboard on wasm and pull in a `web_sys::ClipboardEvent` but I am unsure of the semantics of triggering this callback. The clipboard read is triggered on a click in my program but do I have to attempt to write to the `web_sys::Clipboard` first to trigger the `copy` or such, or put a `click` trigger when I process the click internally? I found information to trigger a `click` has the same problems of user-activation so I believe you can create an invisible text-input under the location of the button and trigger that `.on_click()` to create a valid interaction. 
Also `web_sys::Clipboard::write()` uses a `JsValue` as input, so would that be an `Object` with the `ClipboardItem` promise?

**Reply by daxpedda** [reply]
> I am having trouble converting to `web_sys`. `ClipboardItem` does not have a constructor I can use in `web_sys` as far as I can understand.

That is caused by https://github.com/rustwasm/wasm-bindgen/issues/3587, but you can create a binding yourself:
```rust
#[wasm_bindgen]
extern "C" {
    #[wasm_bindgen(js_name = ClipboardItem, extends = ClipboardItem)]
    type ClipboardItemExt;

    #[wasm_bindgen(constructor, js_class = ClipboardItem)]
    fn new(data: &JsValue, options: &Object) -> ClipboardItemExt;
}
```

> I can use a 'copy' event callback when I initialize the Clipboard on wasm and pull in a `web_sys::ClipboardEvent` but I am unsure of the semantics of triggering this callback. The clipboard read is triggered on a click in my program but do I have to attempt to write to the `web_sys::Clipboard` first to trigger the `copy` or such, or put a `click` trigger when I process the click internally?

The `copy` event is only for receiving `Ctrl-C`, not sure exactly how that's triggered on iOS.
So you should probably use `click`.

> I found information to trigger a `click` has the same problems of user-activation so I believe you can create an invisible text-input under the location of the button and trigger that `.on_click()` to create a valid interaction.

You should test the JS code I showed you above, that would find an answer to that. But I'm pretty sure you won't need an invisible text-input or anything like that.

> Also `web_sys::Clipboard::write()` uses a `JsValue` as input, so would that be an `Object` with the `ClipboardItem` promise?

No, it has to be an array with `ClipboardItem`s, the `ClipboardItem` constructor parameter is an object, which has `Promise` values. I really encourage you to study the JS examples I posted above and try them.

**Reply by miketwenty1** [reply]
@daxpedda I'm happy to try something, but you posted javascript above.

currently I have:
```rust
let item_data = Object::new();

let item_value = Blob::new_with_blob_sequence_and_options(
    &Array::of1(&string_text.into()),
    BlobPropertyBag::new().type_("text/plain"),
).unwrap();

js_sys::Reflect::set(
    &item_data,
    &"text/plain".into(),
    &item_value,
).unwrap();

let item = ClipboardItem::new(&item_data);
let p2 = a.write(&Array::of1(&item));
let result2 = wasm_bindgen_futures::JsFuture::from(p2).await;
```

It's not clear how to convert your js suggestion to rust code.

**Reply by eblack-leaf** [reply]
I tested just the javascript and no luck on iOS for me. It works on every other target I tried.

**Reply by eblack-leaf** [reply]
This also worked in Rust, but not on iOS browsers.

```rust
let js_string = JsValue::from_str(data.as_str());
let js_array = web_sys::js_sys::Array::from_iter(std::iter::once(js_string));
let js_blob = web_sys::Blob::new_with_str_sequence_and_options(
    &js_array,
    &BlobPropertyBag::new().type_("text/plain"),
)
.unwrap();
let inner_promise = wasm_bindgen_futures::js_sys::Promise::resolve(&js_blob);
let js_obj = Object::new();
web_sys::js_sys::Reflect::set(&js_obj, &"text/plain".into(), &inner_promise)
    .unwrap();
let item = ClipboardItemExt::new(&js_obj, &Object::new());
let item_array = web_sys::js_sys::Array::from(item.as_ref());
wasm_bindgen_futures::spawn_local(async move {
    let promise = web_sys::window()
        .expect("window")
        .navigator()
        .clipboard()
        .unwrap()
        .write(&item_array);
    let _message = wasm_bindgen_futures::JsFuture::from(promise).await.ok();
});
```

I don't want to take time away from the amazing work you guys are doing with trait-based approach, so if this is out of the way it can stay on the back-burner.

**Reply by daxpedda** [reply]
> @daxpedda I'm happy to try something, but you posted javascript above.

That's on purpose. The problem we had so far is getting it work by using the Web API, its not a `wasm-bindgen` or Wasm problem. By trying it in JS first we can create a more easily testable and minimal example then we can with Rust.

But adapting your Rust code:
```rust
let item_data = Object::new();

let item_value = Blob::new_with_blob_sequence_and_options(
    &Array::of1(&string_text.into()),
    BlobPropertyBag::new().type_("text/plain"),
)
.unwrap();

js_sys::Reflect::set(
    &item_data,
    &"text/plain".into(),
    &Promise::new(&mut |resolve, _| {
        resolve.call1(&JsValue::UNDEFINED, &item_value).unwrap();
    }),
)
.unwrap();

let item = ClipboardItem::new(&item_data);
let p2 = a.write(&Array::of1(&item));
let result2 = wasm_bindgen_futures::JsFuture::from(p2).await;
```

Just make sure to execute this in the event handler.
Preferably try the JS code if you can.

**Reply by daxpedda** [reply]
> I tested just the javascript and no luck on iOS for me. It works on every other target I tried.

What exactly happened, was there an error message or anything in that regard?

**Reply by daxpedda** [reply]
> ```rust
> let js_string = JsValue::from_str(data.as_str());
> let js_array = web_sys::js_sys::Array::from_iter(std::iter::once(js_string));
> let js_blob = web_sys::Blob::new_with_str_sequence_and_options(
>     &js_array,
>     &BlobPropertyBag::new().type_("text/plain"),
> )
> .unwrap();
> let inner_promise = wasm_bindgen_futures::js_sys::Promise::resolve(&js_blob);
> let js_obj = Object::new();
> web_sys::js_sys::Reflect::set(&js_obj, &"text/plain".into(), &inner_promise)
>     .unwrap();
> let item = ClipboardItemExt::new(&js_obj, &Object::new());
> let item_array = web_sys::js_sys::Array::from(item.as_ref());
> wasm_bindgen_futures::spawn_local(async move {
>     let promise = web_sys::window()
>         .expect("window")
>         .navigator()
>         .clipboard()
>         .unwrap()
>         .write(&item_array);
>     let _message = wasm_bindgen_futures::JsFuture::from(promise).await.ok();
> });
> ```

```rust
let item_array = web_sys::js_sys::Array::from(item.as_ref());
```
This looks incorrect to me, it should be `Array::of1()` instead of `Array::from()`.

**Reply by eblack-leaf** [reply]
The script I used (which worked on other targets)

```javascript
<script type="application/javascript">
    window.addEventListener('click', async () => {
        await navigator.clipboard.write([
            new ClipboardItem({
                'text/plain': new Promise(async (resolve) => resolve(new Blob(['clipboard test'], {type: 'text/plain'}))),
            }),
        ])
    })
</script>
```

Error `Script error. at 0:0`

**Reply by daxpedda** [reply]
That's a terrible error message ... I'm not even sure if that means it isn't working or there is something wrong with the script setup.

**Reply by eblack-leaf** [reply]
If the same script worked on ubuntu firefox/chrome and android firefox/chrome but not on iOS firefox/chrome/safari, are there iOS only considerations for scripts that could help eliminate script setup? 
It is pretty much a blank page except the script.

**Reply by daxpedda** [reply]
By the way, you only need to test one browser on iOS, they are all using the same WebKit engine (basically Safari) underneath.

> ... are there iOS only considerations for scripts that could help eliminate script setup?

I honestly wouldn't know, you will have to play around with it.
Unfortunately I can't test it myself without an iOS device.

**Reply by eblack-leaf** [reply]
Thanks for all the help. Just to note on the same device I tried the `demo` at the bottom of the [web.dev](https://web.dev/articles/async-clipboard) which uses the `Clipboard API` and it does work so I know it can be done, just have to keep plugging away.

**Reply by daxpedda** [reply]
So I got my hands on an iOS device and tried around a bit.
For some reason I couldn't get a `pointerdown` event to fire on `window`, which is weird.
But here is a working example: [JsFiddle](https://jsfiddle.net/6df27omv/4), even works without an additional `Promise`.

Let me know if you have any trouble translating this to Rust or if anything else comes up.

**Reply by eblack-leaf** [reply]
Well it does indeed work in the `jsFiddle` example on the device, but Rust approach does not work. I can only think the user activation is blocked on webkit for programmatic approaches, whereas the javascript example has a visible button for valid interaction.

**Reply by daxpedda** [reply]
No it works on a canvas as well: https://jsfiddle.net/d36sngzo
Tried it on both an iPad and an iPhone.
It should work on every element as long as its not directly the `Window` I assume.
Also tried it in Rust and it works there as well.

If you are still having issues let me know and I'm happy to look at some code.

**Reply by eblack-leaf** [reply]
I am using it here [clipboard.rs](https://github.com/eblack-leaf/foliage/blob/main/foliage_proper/src/clipboard.rs) lines `124 - 166` and `208 - 216`. I would love the feature but I may be not the best `web-sys`-ian. Thanks for checking it out if you have the time.

**Reply by daxpedda** [reply]
So I just took your code and without any modification at all it just worked.
If I click the button I get the text "testing-testing-456" into my clipboard.

I assume what doesn't work is the `write()` function?
I can see that you are going through creating a temporary button and use the `HTMLElement.click()` function on it, that obviously won't work because you need transient activation.

I see that you are using Winit. If you call the `write()` function inside the `WindowEvent::MouseInput` event, then call `Clipboard.write()` it should work.

I will whip up a small example.

**Reply by daxpedda** [reply]
I believe the first click always fails because the canvas is not in focus yet, its probably an added protection to transient activation.
This obviously doesn't happen in a real application, because the very first thing you do is not writing to the clipboard!

Tried this on Desktop, Android (Firefox), iPad and iPhone, all works!

<details>
  <summary><code>Cargo.toml</code></summary>

```toml
[package]
name = "test"
version = "0.0.0"
edition = "2021"
publish = false

[dependencies]
softbuffer = "0.4"
wasm-bindgen-futures = "0.4"
web-sys = { version = "0.3", features = [
    "Blob",
    "BlobPropertyBag",
    "console",
    "Clipboard",
    "ClipboardItem",
    "CssStyleDeclaration",
    "HtmlElement",
    "Navigator",
    "Window",
] }
winit = "0.30"
```

</details>

<details>
  <summary><code>main.rs</code></summary>

```rust
use js_sys::{wasm_bindgen, Array, Object, Reflect};
use softbuffer::{NoDisplayHandle, NoWindowHandle, Surface, SurfaceExtWeb};
use wasm_bindgen::{prelude::wasm_bindgen, JsValue};
use wasm_bindgen_futures::JsFuture;
use web_sys::{console, js_sys, Blob, BlobPropertyBag, Clipboard};
use winit::application::ApplicationHandler;
use winit::error::EventLoopError;
use winit::event::{ElementState, MouseButton, StartCause, Touch, TouchPhase, WindowEvent};
use winit::event_loop::{ActiveEventLoop, EventLoop};
use winit::platform::web::{WindowAttributesExtWebSys, WindowExtWebSys};
use winit::window::{Window, WindowId};

fn main() -> Result<(), EventLoopError> {
    EventLoop::new().unwrap().run_app(&mut App::Inactive)
}

enum App {
    Inactive,
    Active {
        _window: Window,
        surface: Surface<NoDisplayHandle, NoWindowHandle>,
    },
}

impl ApplicationHandler for App {
    fn new_events(&mut self, event_loop: &ActiveEventLoop, cause: StartCause) {
        if let StartCause::Init = cause {
            let window = event_loop
                .create_window(Window::default_attributes().with_append(true))
                .unwrap();

            let canvas = window.canvas().unwrap();
            let style = canvas.style();
            style.set_property("width", "100%").unwrap();
            style.set_property("height", "100%").unwrap();

            let surface = Surface::from_canvas(canvas).unwrap();

            *self = Self::Active {
                _window: window,
                surface,
            };
        }
    }

    fn resumed(&mut self, _: &ActiveEventLoop) {}

    fn window_event(&mut self, _: &ActiveEventLoop, _: WindowId, event: WindowEvent) {
        let App::Active { surface, .. } = self else {
            unreachable!()
        };

        match event {
            WindowEvent::MouseInput {
                state: ElementState::Released,
                button: MouseButton::Left,
                ..
            }
            | WindowEvent::Touch(Touch {
                phase: TouchPhase::Ended,
                ..
            }) => {
                thread_local! {
                    static CLIPBOARD: Clipboard = web_sys::window().unwrap().navigator().clipboard().unwrap();
                    static DATA: Array = {
                        let blob = Blob::new_with_blob_sequence_and_options(
                            &Array::of1(&"test clipboard".into()),
                            BlobPropertyBag::new().type_("text/plain"),
                        )
                        .unwrap();
                        let record = Object::new();
                        Reflect::set(&record, &"text/plain".into(), &blob).unwrap();
                        let item = ClipboardItemExt::new(&record);

                        Array::of1(&item)
                    };
                }

                let promise = CLIPBOARD.with(|clipboard| DATA.with(|data| clipboard.write(data)));

                wasm_bindgen_futures::spawn_local(async move {
                    if let Err(error) = JsFuture::from(promise).await {
                        console::error_2(&"writing to clipboard failed: ".into(), &error);
                    }
                });
            }
            WindowEvent::Resized(size) => surface
                .resize(
                    size.width.try_into().unwrap(),
                    size.height.try_into().unwrap(),
                )
                .unwrap(),
            WindowEvent::RedrawRequested => {
                let mut buffer = surface.buffer_mut().unwrap();
                buffer.fill(0x00ff0000);
                buffer.present().unwrap();
            }
            _ => (),
        }
    }
}

#[wasm_bindgen]
extern "C" {
    type ClipboardItemExt;

    #[wasm_bindgen(constructor, js_class = ClipboardItem)]
    fn new(data: &JsValue) -> ClipboardItemExt;
}
```

</details>
